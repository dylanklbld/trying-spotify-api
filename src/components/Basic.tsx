import styled from "styled-components";

export const SectionTitle = styled.h1`
    font-size: 2em;
    text-align: center;
    font-family: Palatino, 'Palatino Linotype', serif;
    color: white;
`

export const PageContainer = styled.div`
  height: 100%;
  min-height: 100vh;
  width: 100%;
`