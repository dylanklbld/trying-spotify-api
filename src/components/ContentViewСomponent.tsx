import React, { useEffect } from 'react'

import { FetchMoreOptions } from 'react-query/types/core/query'
import GalleryContainer from './Gallery/GalleryContainer'
import ShowMoreButton from './Buttons/ShowMoreButton'
import UpButton from './Buttons/ScrollUpButton'
import {useIsInView} from '../hooks/useIsInView'

type Props = {
    items: Array<any>,
    isLoading: boolean,
    canFetchMore: boolean,
    handleFetchMore: (fetchMoreVariable?: any, options?: FetchMoreOptions | undefined) => Promise<any[] | undefined>,
    renderItem:(item: any) => JSX.Element
}

function ContentViewComponent({ items, isLoading, canFetchMore, handleFetchMore, renderItem  }: Props): React.ReactElement {
    const [renderUpButton, setRenderUpButton] = React.useState(false)
    const [refAnchor, inView] = useIsInView<HTMLDivElement>()

    useEffect(() => {
        setRenderUpButton(!inView)
    }, [inView])
    
    const scrollUp = () => {
        refAnchor?.current?.scrollIntoView({block: "center", behavior: "smooth"})
    }

    return (
            <>
                <div ref={refAnchor} id="scroll-top-anchor" />   
                <GalleryContainer isLoading={isLoading}>
                    {items?.map((item: any) => renderItem(item))}
                </GalleryContainer>
                {renderUpButton && <UpButton onClick={scrollUp} />}
                <ShowMoreButton 
                    onClick={handleFetchMore}
                    isLoading={isLoading}
                    isShown={Boolean(canFetchMore)} 
                />
            </>
        )
}

export default ContentViewComponent