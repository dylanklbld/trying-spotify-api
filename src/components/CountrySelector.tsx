import React, { ReactElement, useState } from 'react'
import styled, { keyframes } from "styled-components";

const Container = styled.div`
  width: 90%;
  margin: 0 auto;
  display: flex;
  flex-flow: row;
  justify-content: center;
  text-align: center;
  font-family: "Courier New", Courier, monospace;
  color: white;
`

const FlagButton = styled.button<{isActive?:boolean}>`
  font-size:2em;
  cursor:pointer;
  padding: 1px 8px;
  border-radius: 3px;
  color: #17323F;
  opacity:${props=>props.isActive ? 0.5 : 1}
`;

type Props = {
   countryMap: any
   setCountry: (countryCode: string | null) => void,
   selected?: string | null
}

const resetText = "ðŸ”„"

const CountrySelector = ({ countryMap, setCountry, selected }: Props):ReactElement => {
    return (<Container>
        {Object.keys(countryMap).map((countryCode)=>{
            return <FlagButton key={countryCode} 
            onClick={()=>setCountry(countryCode)} 
            isActive={selected===countryCode}>
                {countryMap[countryCode]}
            </FlagButton>
        })}
        <button onClick={()=>setCountry(null)}>{resetText}</button>
      </Container>)
}

export default CountrySelector