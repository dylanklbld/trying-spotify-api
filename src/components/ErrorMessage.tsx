import React, { useEffect } from 'react'
import styled, { keyframes } from "styled-components";

import { useHistory } from 'react-router-dom';
import useSessionStatus from '../hooks/useSessionStatus'

const ErrorContainer = styled.div`
  width: 100%;
  top:30%;
  margin: 0 auto;
  display: flex;
  flex-flow: column;
  justify-content: center;
  background:#d77070de;
  font-size: 2em;
  text-align: center;
  font-family: "Courier New", Courier, monospace;
  color: white;
  z-index:1;
  position:sticky;
`
const SessionMessage = styled.div`
  width: 50%;
  margin: 0 auto;
  display: flex;
  flex-flow: row wrap;
  justify-content: center;
`

const Message = styled.div`
  width: 50%;
  margin: 0 auto;
  display: flex;
  flex-flow: row wrap;
  justify-content: center;
`

const Title = styled.h2`
  width: 50%;
  margin: 0 auto;
  display: flex;
  flex-flow: row wrap;
  justify-content: center;
`

const ExtraInfo = styled.p`
  margin: 0 auto;
  display: flex;
  flex-flow: row wrap;
`


const errorTitleText = "Oops, something went wrong"
const sessionExpired = "Session is expired, please go to the start page"

type Props = {
  error: Error
}

const ErrorMessage = ({ error }: Props) => {
  const [sessionIsAlive] = useSessionStatus()
  const { message } = error

  return (<ErrorContainer>
    <Title>{errorTitleText}</Title>
    <Message>{message}</Message>
    {!sessionIsAlive && <>
      <SessionMessage>{sessionExpired}</SessionMessage>
      <a href='/'>start page</a>
    </>}

  </ErrorContainer>)
}

export default ErrorMessage