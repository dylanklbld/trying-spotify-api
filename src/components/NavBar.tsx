import {NavLink, useLocation, useRouteMatch} from 'react-router-dom'

import React from 'react'
import styled  from "styled-components";

const navbarColor = `#3B5765`

const Nav = styled.div`
  display: flex;
  flex-flow: column nowrap;
  align-items: center;
  font-size: 2em;
  text-align: center;
  font-family: Verdana,Geneva,sans-serif; 
  color: white;
`;

const NavLinksWrapper = styled.ul`
    margin: 0;
    padding: 0;
    text-align:center;
    list-style: none;
    list-style-type: none;
    height:100%;
    background: ${navbarColor};
`

const NavLinkStyled = styled(props=><NavLink {...props}/>)`
  text-decoration: none;
  font-size: 1em;
  transition: font-size 0.5s linear, text-shadow 1s;
  color: #09212D;
  &.active {
    font-size: 1.5em;
    color: white;
  }
  &:hover {
      cursor:pointer;
      color: white;
      text-shadow: 2px 2px 6px #0e98de;
  }
`;

const MenuItem = styled.li`
    display:inline-block;
    padding: 20px 20px;
`;

type Props = {
    links: any
}

export default function NavBar({links}: Props): React.ReactElement{
    const match = useRouteMatch()
    const { pathname } = useLocation()

    return (<Nav>
        <NavLinksWrapper>
          {links.map(({link, text}:any) => (
            <MenuItem key={link}>
                <NavLinkStyled to={`${match.path}/${link}`}
                    isActive={()=>[`${match.path}/${link}`].includes(pathname)}
                >
                    {text}
                </NavLinkStyled>
            </MenuItem>
            ))}
        </NavLinksWrapper>
    </Nav>)
}