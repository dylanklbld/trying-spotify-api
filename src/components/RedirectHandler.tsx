import React, { useEffect } from 'react'

import {getHashParameters} from '../helpers/getHashParameters'
import localStorage from '../utils/localStorage'
import { useHistory } from 'react-router-dom';

type Props = {
    location?: any,
}

const defaultUrlPath = '/'
const dashboardUrlPath = '/dashboard'

export const apiParametersStorageKey = 'parameters'
export const expiryTimeStorageKey = 'expiry-time'

function RedirectHanlder({ location }: Props): React.ReactElement {
    const history = useHistory();

    useEffect(() => {
        try {
            const extractedParams = getHashParameters(location?.hash) 
            const expiryTime = new Date().getTime() + extractedParams['expires_in'] * 1000;
            
            localStorage.setItem(apiParametersStorageKey, JSON.stringify(extractedParams))
            localStorage.setItem(expiryTimeStorageKey, expiryTime)
            
            history.push(dashboardUrlPath)
        } catch (error) {
            history.push(defaultUrlPath)
        }
    }, [])

    return <div>Redirecting</div>
}

export default RedirectHanlder