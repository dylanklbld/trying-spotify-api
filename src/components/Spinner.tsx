import styled, { keyframes } from "styled-components";

import React from 'react';

const rotate360 = keyframes`
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
`;

export const Spinner = styled.div<{width?:number, height?:number}>`
  animation: ${rotate360} 1s linear infinite;
  transform: translateZ(0);
  border-top: 2px solid white;
  border-right: 2px solid white;
  border-bottom: 2px solid white;
  border-left: 4px solid #09212D;
  background: transparent;
  width: ${(props) => props.width || 24}px;
  height: ${(props) => props.height || 24}px;
  border-radius: 50%;
  position: absolute;
  z-index: 1;
`;
