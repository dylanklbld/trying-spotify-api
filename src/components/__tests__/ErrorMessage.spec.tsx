import { fireEvent, render, screen } from '@testing-library/react';

import ErrorMessage from '../ErrorMessage';
import React from 'react';
import { BrowserRouter as Router } from 'react-router-dom';
import { expiryTimeStorageKey } from '../../components/RedirectHandler'
import storage from '../../utils/localStorage'

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useRouteMatch: () => ({ path: 'dashboard' }),
    useLocation:()=>({ pathname:'dashboard/path' })
}));

describe('Error Message', () => {
  it('should render', () => {
    const err = new Error('Error occured')
    const currentTime = new Date().getTime();
    storage.setItem(expiryTimeStorageKey, currentTime+5000)
    const { container } = render(<ErrorMessage error={err} />)
   
    expect(container).toMatchSnapshot()
  })
  
  it('should render with session expired message', () => {
    const err = new Error('Error occured')
    storage.setItem(expiryTimeStorageKey, 0)
    const { container } = render(<ErrorMessage error={err} />)
   
    expect(container).toMatchSnapshot()
  })

})