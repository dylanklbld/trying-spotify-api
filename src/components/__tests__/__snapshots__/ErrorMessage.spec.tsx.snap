// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Error Message should render 1`] = `
<div>
  <div
    class="sc-bdfBwQ fZFSNC"
  >
    <h2
      class="sc-hKgILt cHahJl"
    >
      Oops, something went wrong
    </h2>
    <div
      class="sc-dlfnbm kLabmn"
    >
      Error occured
    </div>
  </div>
</div>
`;

exports[`Error Message should render with session expired message 1`] = `
<div>
  <div
    class="sc-bdfBwQ fZFSNC"
  >
    <h2
      class="sc-hKgILt cHahJl"
    >
      Oops, something went wrong
    </h2>
    <div
      class="sc-dlfnbm kLabmn"
    >
      Error occured
    </div>
    <div
      class="sc-gsTCUz cYeAWU"
    >
      Session is expired, please go to the start page
    </div>
    <a
      href="/"
    >
      start page
    </a>
  </div>
</div>
`;
